### PR TITLE
27.50 min record on the GPT-2-medium track (~1% to 2% improvement)

### PR DESCRIPTION
**Pull request by leloykun**
### Changelogs
- Optimized Muon coefficients
- 7150 steps -> 7050 steps (7025 would also work, but would require more sample runs)

image

### Ablations

> Do the gains merely come from reducing the number of training steps?

No.
- At the original 7150 training steps, an optimized 6-step NS coeffs improves validation loss by ~`0.0013`.
- At 7050 training steps, the original 5-step NS gets an average val_loss of `2.9201`.
- At 7050 training steps, an optimized 7-step NS gets an average val_loss of `2.9192`.

> Do the gains merely come from increasing the increasing the number of NS iterations?

No.
- At 7050 training steps, the original 7-step NS gets an average val_loss of `2.9199`

#### Baselines
|  | train steps | val_loss | wallclock (mins) |
| --- | --- | --- | --- |
| Original 5-step NS | 7050 | 2.9207 | 27.47
|  | 7050 | 2.9203 | 27.46
|  | 7050 | 2.9192 | 27.49
|  | 7050 | 2.9200 | 27.47
|  | 7050 | 2.9199 | 27.47
| Original 7-step NS | 7050 | 2.9199 | ~
|  | 7050 | 2.9199 | ~
|  | 7050 | 2.9199 | ~

#### Main results
|  | train steps | val_loss | wallclock (mins) |
| --- | --- | --- | --- |
| Optimized 5-step NS V2 | 7050 | 2.9181 | ~
|  | 7050 | 2.9202 | ~
|  | 7050 | 2.9188 | ~
|  | 7050 | 2.9194 | ~
|  | 7050 | 2.9195 | ~
|  | 7050 | 2.9185 | 27.50
|  | 7050 | 2.9193 | 27.49

mean = 2.9191
p-value (<2.92) = 0.0078

#### Other results
|  | train steps | val_loss | wallclock (mins) |
| --- | --- | --- | --- |
| Optimized 5-step NS | 7050 | 2.9244 | 27.48
|  | 7050 | 2.9514 | 27.52
|  | 7050 | 2.9294 | 27.52
| Optimized 6-step NS | 7150 | 2.9169 | 27.91
| Optimized 6-step NS V2 | 7050 | 2.9185 | 27.52
|  | 7050 | 2.9204 | 27.53
|  | 7050 | 2.9205 | 27.53
| Optimized 7-step NS V1 | 7025 | 2.9208 | 27.50
|  | 7025 | 2.9187 | 27.48
| Optimized 7-step NS V2 | 7050 | 2.9189 | 27.57
|  | 7050 | 2.9201 | 27.56
|  | 7050 | 2.9194 | 27.58
| Optimized 7-step NS V3 | 7050 | 2.9185 | 27.59
|  | 7050 | 2.9205 | 27.59
|  | 7050 | 2.9179 | 27.58
| Optimized 7-step NS V4 | 7050 | 2.9197 | 27.57
|  | 7050 | 2.9203 | 27.58
|  | 7050 | 2.9185 | 27.58
| Optimized 7-step NS V5 | 7050 | 2.9185 | ~
|  | 7050 | 2.9184 | ~
|  | 7050 | 2.9198 | ~
|  | 7050 | 2.9197 | ~
|  | 7050 | 2.9208 | ~
|  | 7050 | 2.9184 | ~
|  | 7050 | 2.9196 | ~
|  | 7050 | 2.9188 | ~
|  | 7050 | 2.9193 | ~
|  | 7050 | 2.9201 | ~

### Optimized coefficients

Optimized 5-step V2:
```python
[
    (4.0848, -6.8946, 2.9270),
    (3.9505, -6.3029, 2.6377),
    (3.7418, -5.5913, 2.3037),
    (2.8769, -3.1427, 1.2046),
    (2.8366, -3.0525, 1.2012),
]
```

---

#### Other coefficients

5-step:
```python
[
    (4.1689, -4.3158, 1.1469),
    (4.1656, -4.3124, 1.1468),
    (4.1153, -4.2443, 1.1290),
    (3.6300, -3.5835, 0.9535),
    (2.8416, -2.5875, 0.7459),
]
```

6-step V2:
```python
[
    (3.8090, -3.9438, 1.1348),
    (3.7890, -3.9226, 1.1336),
    (3.5701, -3.6794, 1.1093),
    (3.4926, -3.5834, 1.0908),
    (3.2071, -3.2823, 1.0752),
    (3.0544, -3.1174, 1.0630),
]
```

7-step V2:
```python
[
  (3.5411, -4.7522, 1.6958),
  (3.0775, -2.4176, 0.5946),
  (3.5138, -3.0216, 0.8168),
  (3.1265, -2.3580, 0.5530),
  (3.5892, -3.3929, 0.9762),
  (3.1026, -2.6731, 0.7442),
  (2.6682, -2.4636, 0.7865),
]
```

7-step V3:
```python
[
  (3.9268, -4.0578, 1.1311),
  (3.8707, -3.9821, 1.1114),
  (3.7325, -3.8442, 1.1117),
  (3.5138, -3.6070, 1.0932),
  (3.3625, -3.4231, 1.0606),
  (3.0690, -3.1200, 1.0510),
  (2.8648, -2.8872, 1.0225),
]
```

7-step V4:
```python
[
    (3.9773, -4.1167, 1.1394),
    (3.8026, -3.9035, 1.1008),
    (3.6792, -3.7796, 1.1005),
    (3.4667, -3.5503, 1.0837),
    (3.3450, -3.4149, 1.0700),
    (3.0610, -3.1217, 1.0606),
    (2.8452, -2.8777, 1.0324),
]
```

Optimized 7-step V5:
```python
[
    (3.8533, -3.9406, 1.0872),
    (3.6963, -3.7832, 1.0869),
    (3.5444, -3.6313, 1.0869),
    (3.3965, -3.4832, 1.0867),
    (3.2532, -3.3399, 1.0866),
    (3.1099, -3.1942, 1.0842),
    (2.7204, -2.6644, 0.9440),
]
```

### More ablation results

1. If we want to increase the steepness, we have to increase the number of steps (or probably increase the degree of the iterator from 5 to e.g. 7). Optimizing the coefficients doesn't really help much at some point.
2. If we want to decrease the "noise", we have to use different coefficients per step.
3. I'm also seeing massive gains early in training which then disappear later on.
  - What seems to matter for early training is steepness of the composite iteration. But later in training, noise reduction matters more. I don't know why this is the case.

image
---
image
---
image
---
image

---

Notes to @KellerJordan (or whoever would test):

- The first run in my machine is almost always 1ms/step slower than the succeeding runs. I have no idea what causes this.
- Adding more precision to the NS coefficients causes a performance degradation. But 4 decimal places seem to suffice.

### CODE REVIEW

**Comment by KellerJordan** [comment]
qq: Do we know how the original 5-step NS does at 7050 steps? Is it even worse than the optimized 5-step one you used?

This will be great as a new record. To accept it, we need statistical significance. Btw, for the previous record (7150 steps), IIRC I did run it a bunch of times and get stat sig.

**Comment by YouJiacheng** [comment]
can you also visualize the composition function (i.e. after the iters)?
also curious how this set of coefs compared to others

**Comment by YouJiacheng** [comment]
Also, I think including "7-step original coefficients" and "22-step original 3rd-order NS" (`1.5 ** 22 = 7481.827643` larger than 7-step V3) as baselines might be informative.

**Comment by KellerJordan** [comment]
Based on the outlier for the optimized 5-step coefficients, there's a chance that this method of finding coefficients leads to instability

I've seen this before - I think what happens is that the coefficients are designed for singular values that stay strictly within the range [0, 1]. But then bf16 means it will sometimes go slightly above 1, in which case things blow up. So it can be a good idea to keep the coefficients such that everything in the range [0, ~1.1] converges

**Comment by leloykun** [comment]
@KellerJordan the latest version I have here is a 5-step NS that performs just as well as the optimized 7-step NS, but 0.01 mins faster

```python
[
    (4.0848, -6.8946, 2.9270),
    (3.9505, -6.3029, 2.6377),
    (3.7418, -5.5913, 2.3037),
    (2.8769, -3.1427, 1.2046),
    (2.8366, -3.0525, 1.2012),
]
```

I've also updated the sample log here to use this.

**Comment by YouJiacheng** [comment]
Oh, the final one is the one that has a peak near 0.
image
